syne_tune/backend/sagemaker_backend/sagemaker_utils.py

```diff
@@ -284,3 +285,144 @@ def map_identifier_limited_length(
+    first_error_message = None
+    while more_objects:
+        response = s3_client.list_objects_v2(
+            **list_objects_kwargs, **continuation_kwargs)
+        # Subdirectories
+        for next_prefix in response.get('CommonPrefixes', []):
+            all_next_prefixes.append(next_prefix['Prefix'])
+        # Objects
+        for source in response.get('Contents', []):
+            ret_msg = action(s3_client, bucket, source['Key'])
+            num_action_calls += 1
+            if ret_msg is None:
+                num_successful_action_calls += 1
+            elif first_error_message is None:
+                first_error_message = ret_msg
+        more_objects = 'NextContinuationToken' in response
+        if more_objects:
+            continuation_kwargs = {
+                'ContinuationToken': response['NextContinuationToken']}
+    # Recursive calls
+    for next_prefix in all_next_prefixes:
+        result = _s3_traverse_recursively(
+            s3_client, action, bucket, prefix=next_prefix)
+        num_action_calls += result['num_action_calls']
+        num_successful_action_calls += result['num_successful_action_calls']
+        if first_error_message is None:
+            first_error_message = result['first_error_message']
+    return dict(
+        num_action_calls=num_action_calls,
+        num_successful_action_calls=num_successful_action_calls,
+        first_error_message=first_error_message)
+
+
+def _split_bucket_prefix(s3_path: str) -> (str, str):
+    assert s3_path[:5] == 's3://', s3_path
+    parts = s3_path[5:].split('/')
+    bucket = parts[0]
+    prefix = '/'.join(parts[1:])
+    if prefix[-1] != '/':
+        prefix += '/'
+    return bucket, prefix
+
+
+def s3_copy_files_recursively(s3_source_path: str, s3_target_path: str) -> dict:
+    """
+    Recursively copies files from `s3_source_path` to `s3_target_path`.
+
+    We return a dict with 'num_action_calls', 'num_successful_action_calls',
+    'first_error_message' (the error message for the first failed `action` call
+    encountered).
+
+    :param s3_source_path:
+    :param s3_target_path:
+    :return: See above
+    """
+    src_bucket, src_prefix = _split_bucket_prefix(s3_source_path)
+    trg_bucket, trg_prefix = _split_bucket_prefix(s3_target_path)
+
+    def copy_action(s3_client, bucket: str, object_key: str) -> Optional[str]:
+        assert object_key.startswith(src_prefix), \
+            f"object_key = {object_key} must start with {src_prefix}"
+        target_key = trg_prefix + object_key[len(src_prefix):]
+        copy_source = dict(Bucket=bucket, Key=object_key)
+        ret_msg = None
+        try:
+            s3_client.copy_object(
+                CopySource=copy_source,
+                Bucket=trg_bucket,
+                Key=target_key)
+            logger.debug(f"Copied s3://{bucket}/{object_key}   to   s3://{trg_bucket}/{target_key}")
+        except ClientError as ex:
+            ret_msg = str(ex)
```

Comment:
If we cannot copy a checkpoint, something is probably wrong and we should probably not try to continue as it would be hard to recover such an error.

Reply:
OK, but this would be decided at a higher level. I'll put that in.

Reply:
Actually, already done it, your previous comment.

For deleting checkpoints, I'll keep the warning, because not being able to delete a checkpoint should not be a reason to fail.

Reply:
I am fine if you prefer to just have a warning, but I believe an error would be more suited.
If the user you specifically asks for the checkpoint to be deleted and they cannot be, this should be an error since we cannot do the operation the user asked.